log4j-core/src/main/java/org/apache/logging/log4j/core/LoggerContext.java

```diff
@@ -92,6 +94,15 @@ public class LoggerContext extends AbstractLifeCycle
     private volatile URI configLocation;
     private Cancellable shutdownCallback;
 
+    /**
+     * The default message factory to use while creating loggers.
+     * <p>
+     * The initial value is only set to avoid nullability.
+     * The actual value is populated by {@link #reloadDefaultMessageFactory()} during (re)configuration.
+     * </p>
+     */
+    private MessageFactory defaultMessageFactory = ParameterizedMessageFactory.INSTANCE;
```

Comment:
Sorry, but this still doesn't make sense to me.

The default message factory is something that is fixed for at least each `LoggerContext` and can only be configured through configuration properties. It doesn't make sense to change default message factory each time a configuration file is loaded.

If we would implement such a feature, users would end up with half the loggers using the old MF and half the loggers using the new MF.

Reply:
How shall I fix it? I don't want to use `Logger.getEffectiveMessageFactory(null)`, since it will fix the value to a particular logger implementation, though the logger is subject to change via `newInstance()`. (Converting a `private static` method to package-local is already a clear sign of abstraction leakage...)

> The default message factory is something that is fixed for at least each `LoggerContext` and can only be configured through configuration properties. It doesn't make sense to change default message factory each time a configuration file is loaded.

A custom `LoggerContext` can change the `Logger` (and hence the `MF`) at will. I don't think _"[an `MF`] can only be configured through configuration properties"_ is correct. This is indeed what our `LC` implementations do, but that doesn't mean `LC` impls. out there in the wild follow the same rule. My point is: `LC` allows at-runtime change of `Logger` impl. (and hence, `MF` impl.), and we should cover this case.

> If we would implement such a feature, users would end up with half the loggers using the old MF and half the loggers using the new MF.

That's right. Is it bad? If users re-configure an `LC` with a new `MF`, shouldn't we be using this new `MF` as the default one?

Reply:
Anyway... I've put the default `MF` behind a `Lazy`: a8007767db6f80278a2045b39bba7faa4094ce4a. Let me know what you think.

Reply:
> How shall I fix it? I don't want to use `Logger.getEffectiveMessageFactory(null)`, since it will fix the value to a particular logger implementation, though the logger is subject to change via `newInstance()`. (Converting a `private static` method to package-local is already a clear sign of abstraction leakage...)

IMHO, `LoggerContext#newInstance(LoggerContext, String, MessageFactory)` should not be called with `null` parameters, that's all. Then we have no doubt which MF is used by the `Logger`.

> > The default message factory is something that is fixed for at least each `LoggerContext` and can only be configured through configuration properties. It doesn't make sense to change default message factory each time a configuration file is loaded.
> 
> A custom `LoggerContext` can change the `Logger` (and hence the `MF`) at will. I don't think _"[an `MF`] can only be configured through configuration properties"_ is correct. This is indeed what our `LC` implementations do, but that doesn't mean `LC` impls. out there in the wild follow the same rule. My point is: `LC` allows at-runtime change of `Logger` impl. (and hence, `MF` impl.), and we should cover this case.

Let me be more precise: our `LoggerContext` implementations use Log4j configuration properties to determine the MF. Another `LC` implementation can do whatever it wants, as long as it doesn't pass `null` to the `LoggerRegistry`.

Reply:
> IMHO, `LoggerContext#newInstance(LoggerContext, String, MessageFactory)` should not be called with `null` parameters, that's all. Then we have no doubt which MF is used by the `Logger`.

Are you implying throwing an exception or something when the `MF` of `getLogger(String,MF)` is null? How shall we prevent this from happening?

> > > The default message factory is something that is fixed for at least each `LoggerContext` and can only be configured through configuration properties. It doesn't make sense to change default message factory each time a configuration file is loaded.
> > 
> > 
> > A custom `LoggerContext` can change the `Logger` (and hence the `MF`) at will. I don't think _"[an `MF`] can only be configured through configuration properties"_ is correct. This is indeed what our `LC` implementations do, but that doesn't mean `LC` impls. out there in the wild follow the same rule. My point is: `LC` allows at-runtime change of `Logger` impl. (and hence, `MF` impl.), and we should cover this case.
> 
> Let me be more precise: our `LoggerContext` implementations use Log4j configuration properties to determine the MF. Another `LC` implementation can do whatever it wants, as long as it doesn't pass `null` to the `LoggerRegistry`.

I don't think this would work for 3rd party classes extending from `LC`. You give them the impression that they have control over their choice of `Logger`, (and hence, the `MF`), but under the hood, you just ignore it.

Reply:
> > IMHO, `LoggerContext#newInstance(LoggerContext, String, MessageFactory)` should not be called with `null` parameters, that's all. Then we have no doubt which MF is used by the `Logger`.
> 
> Are you implying throwing an exception or something when the `MF` of `getLogger(String,MF)` is null? How shall we prevent this from happening?

Let me remark, that nothing in the Javadoc allows users to call `getLogger(String, MF)` with a `null` second parameter.
The Javadoc didn't explicitly allow calling the `coreLogger` constructor with a `null` `MF` parameter either, until 2.23.0.
Of course we should protect against such calls by implicitly using the message factory computed using configuration properties.

> I don't think this would work for 3rd party classes extending from `LC`. You give them the impression that they have control over their choice of `Logger`, (and hence, the `MF`), but under the hood, you just ignore it.

`LC` implementations can override the `getLogger(String)` method to provide a different MF.
I really don't see a practical use case, where someone would want to do that.

Reply:
Have you seen the `Lazy` impl. committed? WDYT? If you don't like it, what is your suggestion besides using `Logger.getEffectiveMessageFactory(null)`?

Reply:
Fixed in c1dff86c087c9e4ae5ee7e68ad16385d02dc894e as you wanted. (For the record, I don't agree with this direction, but there are more pressing issues than discussing this detail.)